Fix folderlist on Windows-based web servers

mgilkes opened the pull request:
Pull Request for Issue # .

### Summary of Changes
On line 219, the root folder path is trimmed on the end of both back and forward slashes, and the resulting folder is trimmed of both back and forward slashes.

### Testing Instructions
The changes address the issue of folderlist on Windows-based web servers. You can test this on IIS or WAMP, just access any module (or plugin) that uses folderlist, and you will see "C:\wamp\www\/images" or "C:\inetpub\wwwroot\/images", when it should only show as "images".

Since there are no built-in extensions that use folder list, you can add the following field to any module's xml file to test it:

`<field name="test_field" type="folderlist" default="images" label="TEST_LABEL" directory="" description="TEST_DESC" hide_none="true" hide_default="true" exclude="^[Aa]dministrator$|^[Cc]ache$|^[Cc]omponents$|^[Cc]li$|^[Ii]ncludes$|^[Ll]anguage$|^[Ll]ibraries$|^[Ll]ogs$|^[Mm]odules$|^[Pp]lugins$|^[Tt]emplates$|^[Xx]mlrpc$" />`


### Documentation Changes Required

On non-windows-based web servers, the current code works fine, but on windows-based web servers, like IIS or WAMP, the folderlist options will look like "C:\wamp\www\/images" or "C:\inetpub\wwwroot\/images", when it should only show as "images". This messes up any extension that uses folderlist on windows-based web servers. 

I propose to change line 219 from:
`$folder = trim(str_replace($path,  '', $folder), '/');`
to:
`$folder = trim(str_replace(rtrim($path, '/\\'), '', $folder), '/\\');`

This essentially accommodates Windows systems to make the folderlist work as it should.

mgilkes committed:
Fix folderlist on Windows-based web servers

On non-windows-based web servers, the current code works fine, but on windows-based web servers, like IIS or WAMP, the folderlist options will look like "C:\wamp\www\/images" or "C:\inetpub\wwwroot\/images", when it should only show as "images". This messes up any extension that uses folderlist on windows-based web servers. 

I propose to change line 219 from:
$folder = trim(str_replace($path,  '', $folder), '/');
to:
$folder = trim(str_replace(rtrim($path, '/\\'), '', $folder), '/\\');

This essentially accommodates Windows systems to make the folderlist work as it should.

brianteeman commented:
Is this the same as #11524 and #11288 ??

ggppdk commented:
This is more proper fix for folderlist than #11288,
since it catches all cases
plus it does not fail the tests units

@mgilkes 
Can you also include a fix for:
```php
JFolder::folders(...);
```

**Returning paths like**:
D:\xampp\htdocs\devel\joomla-cms2\images\sampledata\/fruitshop
D:\xampp\htdocs\devel\joomla-cms2\images\sampledata\/parks


https://github.com/joomla/joomla-cms/blob/staging/libraries/joomla/filesystem/folder.php#L619
should be:
```php
$fullpath = rtrim($path, '/\\') . DIRECTORY_SEPARATOR  . $file;
```

mgilkes commented:
This issue still exists in Joomla 3.6.4. My solution is different from those proposed. One of the problems with the current code is that the $path variable string actually has a trailing slash, which is why the folder root path is not removed by the str_replace on line 219. It should be trimmed.

ggppdk commented:
I have tested this item :white_check_mark: successfully on 972fa749279d8081a6dd0c9ade86e2d44b74619d<br /><br />There is an improvement that i suggested above in JFolder::folders(...); to return clean paths, can be in different PR<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12841">issues.joomla.org/tracker/joomla-cms/12841</a>.</sub>

ggppdk commented:
I have tested this item :red_circle: unsuccessfully on 972fa749279d8081a6dd0c9ade86e2d44b74619d<br /><br />This PR as the other PR for folderlist does not work in all cases

If the original folder has "/" then JFolder will "clean" it thus replace the "/" with "\" (windows seperator)
in that case the str_replace to remove the given start path will fail<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12841">issues.joomla.org/tracker/joomla-cms/12841</a>.</sub>

mgilkes commented:
@ggppdk I don't quite understand your comment. When you say "original folder" are you referring to the value of $path or $folder (just before line 219)?

Because $path is the root path of the folder list, and $folder is the full path of a subfolder in the folder list, the purpose of line 219 is to remove the root path from the full path, and leave only the subfolder, without any slashes before or after it. Could you better clarify the use case in which:

`$folder = trim(str_replace(rtrim($path, '/\\'), '', $folder), '/\\');`

would fail to provide the subfolder name? Thanks.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12841">issues.joomla.org/tracker/joomla-cms/12841</a>.</sub>

ggppdk commented:
JFolder::folders($path,  ...

Returns:
$folder LIKE: **CLEANED($path)** /somefolder1/someofolder2
- thus **CLEANED($path)** will contain  current server directory separator, thus it will contain \

Thus
$folder = trim(str_replace(rtrim($path, '/\\'), '', $folder), '/\\');
will fail if $path has any inner / and you are in windows

e.g. this will fail
```
<field ... type="folderlist" ...  directory="images/sampledata"
```

but you PR will fail even fail with:
```
<field ... type="folderlist" ...  directory="images" 
```

because folderlist does:
```php
if (!is_dir($path))
{
	$path = JPATH_ROOT . '/' . $path;
}
```

In any case please test this PR:  https://github.com/joomla/joomla-cms/pull/11288
(it has conflicts and needs to be rebased) , it would have been best if you had searched and post at that PR, but i guess you missed it

mgilkes committed:
Cleaned the $path variable to fix mixed slashes

I needed to clean the $path contents since it will be used later when trying to remove it from the subfolder's full path. It is cleaned inside of JFolder::folders, but only in the scope of that function. It needs to be cleaned before it is used by JFolder::folder

mgilkes commented:
Thanks for the clarification. You are right. I think the best thing to do is to clean the $path variable before using it to find the subfolders with JFolder. See my update: [f73febd](https://github.com/joomla/joomla-cms/pull/12841/commits/f73febd1dcc3a122ed70a47b7bc29834b992ed24 "f73febd")<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12841">issues.joomla.org/tracker/joomla-cms/12841</a>.</sub>

wilsonge commented:
Closing this in favour of https://github.com/joomla/joomla-cms/pull/14139 as we had 4 different PR's for the same issue :)
